### ISSUE TITLE
Compile time query support

### ISSUE THREAD

**Issue by Timmmm** [rusqlite/rusqlite#801]
It would be cool if this crate supported some of the [compile time query features that SQLx has](https://github.com/launchbadge/sqlx#compile-time-verification).

**Comment by gwenn** [comment]
We should be able to validate query like [sql!](https://github.com/sfackler/rust-postgres-macros#sql).
But due to [dynamic typing](https://sqlite.org/datatype3.html), we may not be able to deduce the correct type...

**Comment by Timmmm** [comment]
Syntax checking would be a great start. I would love it if Rusqlite could also deduce the correct type from the schema too (SQLite's "it's a feature" is a feeble excuse at best, downright idiotic at worst). But I'll settle for compile time syntax checking!

**Comment by thomcc** [comment]
https://github.com/gwenn/lemon-rs/ contains a (slightly out of date) SQLite parser. If somehow we could verify the correct parameter names/counts that would be a huge start.

That said, `sqlx` expects you to use some sort of "live development database", which it uses to get type information. This is kind of unidiomatic for SQLite in my experience. I'd love us to have some smarts around schema for this sort of thing.

I agree that the dynamic typing is a concern, but I think as long as we don't force checking (and instead just offer a checked API) it wouldn't be a problem.

I do think this should all be optional though, since proc macros add quite a bit to compile time, and currently we don't use any.

**Comment by Timmmm** [comment]
Yes I think having a live database is rather overkill. I would rather specify the schema I expect in the source code. For example in SQLite you can do this:

```
SELECT 
  m.name as table_name,
  p.cid as column_id,
  p.name as column_name,
  p.type as column_type,
  p.`notnull` as column_notnull
FROM 
  sqlite_master AS m
JOIN 
  pragma_table_info(m.name) AS p
ORDER BY 
  m.name, 
  p.cid
```

Which gives you a table something like this:

![image](https://user-images.githubusercontent.com/376842/95086310-400aef00-0718-11eb-979e-060f5d1ff0ad.png)

How hard would it be to allow copy/pasting into a macro something like this?

```
define_schema!(myschema,
compilation_trace|0|id|INTEGER|0
compilation_trace|1|cat|TEXT|1
compilation_trace|2|dur|INTEGER|1
compilation_trace|3|name|TEXT|1
compilation_trace|4|ph|TEXT|1
compilation_trace|5|pid|INTEGER|1
compilation_trace|6|tid|INTEGER|1
compilation_trace|7|ts|INTEGER|1
compute_sets|0|compute_set|INTEGER|0
compute_sets|1|name|TEXT|1
);

let statement = sql!(myschema, "SELECT cat FROM compilation_trace");
```

**Comment by gwenn** [comment]
I am afraid that currenlty `lemon-rs` is not a good candidate mainly because it may panic while parsing but it should be up to date with last stable version (ie 3.33.0).
And `pragma_table_info` is not useful when you have [expressions](https://sqlite.org/lang_expr.html).
Finally, we don't need a database but only a SQL schema (like Jetbrains Database plugin) except for virtual table and pragma.

So we need:
- [ ] SQL(ite) parser for `CREATE TABLE ...`, `CREATE VIEW ...` and `SELECT ...`
- [ ] https://sqlite.org/pragma.html#pragma_function_list to retrieve the list of supported functions,
- [ ] https://sqlite.org/pragma.html#pragma_pragma_list to retrieve the list of supported pragma,
- [ ] minimal SQL compiler to check that all referenced tables, views, columns, functions, pragma are OK
- [ ] ...

**Comment by gwenn** [comment]
[sqlx](https://github.com/launchbadge/sqlx/blob/master/sqlx-core/src/sqlite/connection/explain.rs) parses `EXPLAIN` output.

**Comment by Timmmm** [comment]
By the way, even if there isn't support for validating queries against a schema, just checking the SQL syntax would be good. I keep getting extra commas by writing queries like this:

```
query("SELECT
  foo,
  bar,
  baz
FROM table");
```

And because SQL doesn't support trailing commas if you delete `baz` it results in a runtime syntax error, which could presumably be checked at compile time with nothing other than a SQL parser.

**Comment by gwenn** [comment]
See https://github.com/gwenn/rusqlite/tree/rusqlite-macros/rusqlite-macros

Todo:
* publish `lemon-rs` on crates.io
* find a way to transform parse error location into `Span`
* check SQL statements against database schema.
* should `rusqlite-macros` be in a dedicated repository ?
* ...

**Comment by thomcc** [comment]
Wow!!! Haven't looked yet but that's awesome!

> should `rusqlite-macros` be in a dedicated repository?

I don't think so, since it's possible we may want to change an API in rusqlite, which also impacts the macros.

**Comment by gwenn** [comment]
The (easy) next step would be to to validate the number of place holders / parameters / variables (and their optional names).
But this is possible only if they are specified inline.
And SQLite is fine when any parameter is not specified (`NULL` is used instead).
Are you ok if we raise an error only when we are sure (it doesn't seem possible to raise warning in stable, see `proc_macro_diagnostic` feature) ?

**Comment by gwenn** [comment]
Ok,
I am going to publish [lemon-rs](https://github.com/gwenn/lemon-rs).
But I want to give the crate a different name because lemon is only the parser generator.
And I can't use [sqlite-parser](https://crates.io/crates/sqlite_parser) anymore.
Are you okay with `sqlite3-parser` ?

**Comment by thomcc** [comment]
> https://crates.io/crates/sqlite_parser

Oof, missed by just a bit.

> Are you okay with sqlite3-parser or lemon-sql ?

sqlite3-parser seems clearer to me.

**Comment by gwenn** [comment]
Ok,
we should be able to check parameters now:
https://github.com/gwenn/lemon-rs/blob/master/src/lexer/sql/test.rs
I am not sure mixed parameters are count properly though:
https://sqlite.org/lang_expr.html#parameters
https://github.com/gwenn/lemon-rs/blob/master/src/parser/ast/mod.rs#L75-L85

**Comment by gwenn** [comment]
Just for reference: https://github.com/google/cpp-from-the-sky-down/tree/master/rust (full SQLite parser ?)

**Comment by gwenn** [comment]
@Cloudef see #1346 for syntax error (but not against a DB schema) and placeholders names / count.

**Comment by Cloudef** [comment]
@gwenn nice, if it was possible for it to also do `EXPLAIN QUERY PLAN {}` on a memory db which you can fill with schema as I do above with `execute_batch(&String::from_utf8_lossy(include_bytes!("src/db.sql")))?;` then that would also check typos in table names etc.

**Comment by gwenn** [comment]
As far as I know, `EXPLAIN QUERY PLAN` doesn't check placeholders.
But indeed `EXPLAIN QUERY PLAN` checks query against DB schema.
